nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/AbstractJsonQueryElasticsearch.java

```diff
@@ -0,0 +1,293 @@
+        if (aggregations != null && !aggregations.isEmpty()) {
+            final List<FlowFile> aggsFlowFiles = new ArrayList<>();
+            if (splitUpAggregations.equals(SPLIT_UP_YES.getValue())) {
+                int aggCount = 0;
+                for (final Map.Entry<String, Object> agg : aggregations.entrySet()) {
+                    final FlowFile aggFlowFile = createChildFlowFile(session, parent);
+                    final String aggJson = mapper.writeValueAsString(agg.getValue());
+                    aggsFlowFiles.add(writeAggregationFlowFileContents(agg.getKey(), ++aggCount, aggJson, session, aggFlowFile, attributes));
+                }
+            } else {
+                final FlowFile aggFlowFile = createChildFlowFile(session, parent);
+                final String json = mapper.writeValueAsString(aggregations);
+                aggsFlowFiles.add(writeAggregationFlowFileContents(null, null, json, session, aggFlowFile, attributes));
+            }
+
+            if (!aggsFlowFiles.isEmpty()) {
+                session.transfer(aggsFlowFiles, REL_AGGREGATIONS);
+                aggsFlowFiles.forEach(ff -> session.getProvenanceReporter().receive(ff, transitUri, stopWatch.getElapsed(TimeUnit.MILLISECONDS)));
+            }
+        }
+    }
+
+    private FlowFile writeHitFlowFile(final int count, final String json, final ProcessSession session,
+                                      final FlowFile hitFlowFile, final Map<String, String> attributes) {
+        final FlowFile ff = session.write(hitFlowFile, out -> out.write(json.getBytes()));
+        attributes.put("hit.count", Integer.toString(count));
+
+        return session.putAllAttributes(ff, attributes);
+    }
+
+    List<FlowFile> handleHits(final List<Map<String, Object>> hits, final Q queryJsonParameters, final ProcessSession session,
+                              final FlowFile parent, final Map<String, String> attributes, final List<FlowFile> hitsFlowFiles,
+                              final String transitUri, final StopWatch stopWatch) throws IOException {
+        if (hits != null && !hits.isEmpty()) {
+            if (SPLIT_UP_YES.getValue().equals(splitUpHits)) {
+                for (final Map<String, Object> hit : hits) {
+                    final FlowFile hitFlowFile = createChildFlowFile(session, parent);
+                    final String json = mapper.writeValueAsString(hit);
+                    hitsFlowFiles.add(writeHitFlowFile(1, json, session, hitFlowFile, attributes));
+                }
+            } else {
+                final FlowFile hitFlowFile = createChildFlowFile(session, parent);
+                final String json = mapper.writeValueAsString(hits);
+                hitsFlowFiles.add(writeHitFlowFile(hits.size(), json, session, hitFlowFile, attributes));
+            }
+        }
+
+        // output any results
+        if (!hitsFlowFiles.isEmpty()) {
```

Comment:
It will always be empty for processing non-combined results of paginated queries.

It's used for combining the hits between pages of results with the `List<FlowFile>` being returned by `handleHits` => `handleResponse` => `doQuery` to be sent back through that chain on the next iteration through `doQuery` so that the existing `FlowFile` can be reused to append new hits into.

I agree this is a little confusing when reading the `AbstractJsonQueryElasticsearch#handleHits` method code on its own - maybe some comments here would be useful? Could alternatively consider refactoring the method into separate methods for the `combined` vs. `non-combined` result handling and not return anything for those cases - I'm not sure how much complexity that might add (or aleviate) as the whole chain of method calls will be impacted, so will give it a quick try and then resort to some comments otherwise

Reply:
I think just method comments in both AbstractJsonQueryElasticsearch and AbstractPaginatedJsonQueryElasticsearch should suffice.